include/libgit4cpp/GitRepository.h

```diff
@@ -170,15 +171,49 @@ class GitRepository
     void reset(unsigned int nr_of_commits);
 
     /**
-     * Push commits to the repository.
-     * \param addr Address of the remote git repository host
-    */
-    void push();
+     * Add a new git remote with the specified name and URL to the repository.
+     * \returns a Remote object representing the newly added remote
+     * \exception Error is thrown if the remote cannot be added, for instance because it
+     *            already exists.
+     *
+     * \code{.cpp}
+     * GitRepository repo{ "/path/to/repo" };
+     * repo.add_remote("origin", "https://gitlab.com/a/b.git");
+     * repo.add_remote("upstream", "file:///path/to/upstream/repo");
+     * \endcode
+     */
+    Remote add_remote(const std::string& remote_name, const std::string& url);
+
+    /**
+     * Look up a git remote by name in the repository.
+     * \returns a Remote object if the remote exists, or an empty optional otherwise.
+     */
+    gul14::optional<Remote> get_remote(const std::string& remote_name) const;
 
+    /**
+     * Return a list of all configured git remotes.
```

Comment:
Does that distinction matter? Yes, we spend a tiny amount of extra work to retrieve the URLs as well and to make them easily available, but... why not? Sure, we can return a list of remote names and make the user retrieve the repository pointer, and then they can retrieve the URL if they like. In that case we do not need the `Remote` class at all. It is just an attempt to raise the level of abstraction.

Reply:
At least it should be called `get_all_remotes()` or similar.

And the 'tiny amount of extra work' is what makes you use the less readable `gul::cat()` instead of much more readable `{fmt}` or similar. "But cat saves a tiny amount of work", yes right, which is of course very important when an error message is constructed which is not even in the vicinity of a hot path.

I stumbled about a related thing with `SequenceManager::list_sequences()`. Once I put a debug message in the inside and was astonished how often it is called in an ordinary server.
Example:
* Want to remove a Sequence: `remove_sequence(UniqueId)`
* Calls `list_sequences()`
* Checks every single directory entry if it is a Sequence and constructs many many strings
* Now `find_sequence_on_disk(unique_id, sequences)`
* Walk trough the vector `list_sequences()` returned and search for the one with the right UniqueID to find the path
* Now that Sequence is removed

This does a lot extra work which is never used. Instead a `find_sequence_on_disk(unique_id)` would be more to the point because it could search much more direct. Or a `list_sequences(glob)`.

> In that case we do not need the Remote class at all

Yes. The need for that class is indeed a question. The previous approach with a `remote_` data member in the `GitSequence` was not bad either. The only limitation was that it was not a vector but only one single remote, which can be limiting (but then, that could also be changed without a new class `Remote`).

Anyhow, a `Remote` object by itself is not too helpful as it does not even remember which repo it belonged to. If I want to handle multiple repos and their remotes that becomes a hurdle. I as user need to keep the association myself somewhere. I would not need to do that if the remote is an integral part of the repository object.

Ok.

I try to see where and how `list_remotes()` would be used.
If I want to list the existing remotes I would just need the names as string.
If I want to modify one specific remote I already have the target-remote name somehow and can select that remote directly.
If I want to push to a specific remote I have the target-remote name already; checking if the name exists does not need all remotes as `Remote`, this can be decided on the `push()` call specifically for that name (compare Sequence example above), and/or I can live with a vector of names.

And then, why do we have the members AT ALL?

```c++
class Remote
{
    /// Return a non-owning pointer to the underlying git remote object.
    git_remote* get() const { return remote_.get(); }

    const std::string& get_name() const { return name_; }
    const std::string& get_url() const { return url_; }

private:
    std::string name_;
    std::string url_;
    LibGitRemote remote_{ nullptr, git_remote_free };
```
versus
```c++
class Remote
{
    /// Return a non-owning pointer to the underlying git remote object.
    git_remote* get() const { return remote_.get(); }

    std::string get_name() const { return git_remote_name(remote_.get()); }
    std::string get_url() const { return git_remote_url(remote_.get()); }

private:
    LibGitRemote remote_{ nullptr, git_remote_free };
```

I do not see why we need the members at all. And with just the `LibGitRemote` object as member, `Remote` is kind of ... useless? :thinking: 

When I examine where `Remote` is used, e.g. `push()` only the `LibGitRemote` member is utilized at all.

Reply:
Just for comparison, what other people try to put into the `Remote` object:

```c++
namespace git
{
    struct Remote
    {
        const char * url()      const;
        const char * pushurl()  const;

        struct FetchCallbacks
        {
            virtual void update_tips(char const * refname, git_oid const & a, git_oid const & b) {}
            virtual void sideband_progress(char const * str, int len) {}
            virtual void transfer_progress(git_indexer_progress const &) {}

            virtual git_credential* acquire_cred(const char * url, const char * username_from_url, unsigned int allowed_types) = 0;
        };

        void fetch(FetchCallbacks &, char const * reflog_message = nullptr);

    private:
        struct Destroy { void operator() (git_remote*) const; };

        explicit Remote(git_remote * remote) : remote_(remote) {}

        std::unique_ptr<git_remote, Destroy> remote_;
    };
}
```

https://github.com/AndreyG/libgit2cpp/blob/master/include/git2cpp/remote.h